receiver/receivertest/contract_checker.go

```diff
@@ -0,0 +1,382 @@
+	require.NoError(params.T, err)
+
+	err = receiver.Start(ctx, componenttest.NewNopHost())
+	require.NoError(params.T, err)
+
+	// Begin generating data to the receiver.
+
+	var generatedIds IDSet
+	var generatedIndex int64
+	var mux sync.Mutex
+	var wg sync.WaitGroup
+
+	const concurrency = 4
+
+	// Create concurrent goroutines that use the generator.
+	// The total number of generator calls will be equal to params.GenerateCount.
+
+	for j := 0; j < concurrency; j++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for {
+				if atomic.AddInt64(&generatedIndex, 1) >= int64(params.GenerateCount) {
+					// Generated as many as was requested. We are done.
+					return
+				}
+
+				ids := params.Generator.Generate()
+
+				mux.Lock()
+				duplicates := generatedIds.Merge(ids)
+				mux.Unlock()
+
+				// Check that the generator works correctly. There may not be any duplicates in the
+				// generated data set.
+				require.Empty(params.T, duplicates)
+			}
+		}()
+	}
+
+	// Wait until all generator goroutines are done.
+	wg.Wait()
+
+	// Wait until all data is seen by the consumer.
+	assert.Eventually(params.T, func() bool {
+		// Calculate the union of accepted and dropped data.
+		acceptedAndDropped, duplicates := consumer.acceptedIds.Union(consumer.droppedIds)
+		if len(duplicates) != 0 {
+			assert.Failf(params.T, "found duplicate elements in received and dropped data", "keys=%v", duplicates)
+		}
+		// Compare accepted+dropped with generated. Once they are equal it means all data is seen by the consumer.
+		missingInOther, onlyInOther := generatedIds.Compare(acceptedAndDropped)
+		return len(missingInOther) == 0 && len(onlyInOther) == 0
+	}, 5*time.Second, 10*time.Millisecond)
+
+	// Do some final checks. Need the union of accepted and dropped data again.
+	acceptedAndDropped, duplicates := consumer.acceptedIds.Union(consumer.droppedIds)
+	if len(duplicates) != 0 {
+		assert.Failf(params.T, "found duplicate elements in accepted and dropped data", "keys=%v", duplicates)
+	}
+
+	// Make sure generated and accepted+dropped are exactly the same.
+
+	missingInOther, onlyInOther := generatedIds.Compare(acceptedAndDropped)
+	if len(missingInOther) != 0 {
+		assert.Failf(params.T, "found elements sent that were not delivered", "keys=%v", missingInOther)
+	}
+	if len(onlyInOther) != 0 {
+		assert.Failf(params.T, "found elements in accepted and dropped data that was never sent", "keys=%v", onlyInOther)
+	}
+
+	err = receiver.Shutdown(ctx)
+	assert.NoError(params.T, err)
+
+	// Print some stats to help debug test failures.
+	fmt.Printf(
+		"Sent %d, accepted=%d, expected dropped=%d, non-permanent errors retried=%d\n",
+		len(generatedIds.m),
+		len(consumer.acceptedIds.m),
+		len(consumer.droppedIds.m),
+		consumer.nonPermanentFailures,
+	)
+}
+
+// IDSet is a set of unique ids of data elements used in the test (logs, spans or metric data points).
+type IDSet struct {
+	m map[UniqueIDAttrDataType]bool
+}
```

Comment:
nit: Can this just be an alias?
```suggestion
type IDSet map[UniqueIDAttrDataType]bool
```

Reply:
Done.